CompletionHandler cannot be called from arbitrary thread

kkinnunen-apple opened the pull request:
#### 62b32a176f306013c753c4c506237d3967eff21c
<pre>
CompletionHandler cannot be called from arbitrary thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=246271">https://bugs.webkit.org/show_bug.cgi?id=246271</a>
rdar://problem/100969234

Reviewed by Chris Dumez.

Add CompletionHandlerCallThread::AnyThread so that one can pass
completion handlers that can be called from an arbitrary thread.

Implement with ThreadLikeAssertion, so that the thread check is
exact.

Add symbolic thread names for specifying the ThreadAssertions.

Make the symbolic thread names as their own types (AnyThreadLike,
... ) so that most of the ThreadLikeAssertion constructing callchains
can be constexpr. Also in future APIs, these can be used in the type
definititions instead of relying on assertions. Example:
CompletionHandler&lt;void(), AnyThreadLike&gt; (if this turns out
to make sense).

* Source/WTF/wtf/CompletionHandler.h:
(WTF::CompletionHandler&lt;Out):
(WTF::CompletionHandlerWithFinalizer&lt;Out):
* Source/WTF/wtf/ThreadAssertions.h:
(WTF::MainThreadLike::operator uint32_t const):
(WTF::CurrentThreadLike::operator uint32_t const):
(WTF::AnyThreadLike::operator uint32_t const):
(WTF::NoneThreadLike::operator uint32_t const):
(WTF::ThreadLikeAssertion::ThreadLikeAssertion):
(WTF::ThreadLikeAssertion::operator=):
(WTF::ThreadLikeAssertion::isCurrent const):
(WTF::WTF_ASSERTS_ACQUIRED_CAPABILITY):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/CompletionHandlerTests.cpp: Added.
(TestWebKitAPI::TEST_F):
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WTF/RunLoop.cpp:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Utilities.h:

Canonical link: <a href="https://commits.webkit.org/255576@main">https://commits.webkit.org/255576@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/112d08e7cbdda906d68befdf87182d19c2f7eb3b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92235 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1465 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22822 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102021 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/162332 "Reverted pull request changes (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96235 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1462 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29858 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84675 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98186 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97892 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/954 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78756 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27905 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82871 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82534 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70940 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/83650 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36279 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16493 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/78647 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34035 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17658 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27210 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3862 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37908 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40289 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/81269 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39807 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36799 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18087 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/ThreadAssertions.h:
```diff
@@ -31,6 +31,27 @@
 class ThreadLikeAssertion;
+WTF_EXPORT_PRIVATE bool isMainThread();
+
+struct MainThreadLike {
```
We'll likely need a MainRunLoopThreadLike too at some point. We shouldn't be "MainThread" checks in WebKit2 code, and it would be a bug in the UIProcess because of iOS applications using both WK1 and WK2 (and thus having a WebThread in the UIProcess).

cdumez commented on Source/WTF/wtf/CompletionHandler.h:
```diff
@@ -33,7 +33,12 @@
+public:
+    static inline constexpr auto ConstructionThread = currentThreadLike;
+    static inline constexpr auto MainThread = mainThreadLike;
+    static inline constexpr auto AnyThread = anyThreadLike;
```
As I said offline, I understand you have a particular need for this coming up. However, I hope usage of this doesn't get expanded without a good reason because I don't think this is a good pattern in general.

cdumez reviewed (approved):
LGTM

kkinnunen-apple committed:
CompletionHandler cannot be called from arbitrary thread
https://bugs.webkit.org/show_bug.cgi?id=246271
rdar://problem/100969234

Reviewed by Chris Dumez.

Add CompletionHandlerCallThread::AnyThread so that one can pass
completion handlers that can be called from an arbitrary thread.

Implement with ThreadLikeAssertion, so that the thread check is
exact.

Add symbolic thread names for specifying the ThreadAssertions.

Make the symbolic thread names as their own types (AnyThreadLike,
... ) so that most of the ThreadLikeAssertion constructing callchains
can be constexpr. Also in future APIs, these can be used in the type
definititions instead of relying on assertions. Example:
CompletionHandler<void(), AnyThreadLike> (if this turns out
to make sense).

* Source/WTF/wtf/CompletionHandler.h:
(WTF::CompletionHandler<Out):
(WTF::CompletionHandlerWithFinalizer<Out):
* Source/WTF/wtf/ThreadAssertions.h:
(WTF::MainThreadLike::operator uint32_t const):
(WTF::CurrentThreadLike::operator uint32_t const):
(WTF::AnyThreadLike::operator uint32_t const):
(WTF::NoneThreadLike::operator uint32_t const):
(WTF::ThreadLikeAssertion::ThreadLikeAssertion):
(WTF::ThreadLikeAssertion::operator=):
(WTF::ThreadLikeAssertion::isCurrent const):
(WTF::WTF_ASSERTS_ACQUIRED_CAPABILITY):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/CompletionHandlerTests.cpp: Added.
(TestWebKitAPI::TEST_F):
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WTF/RunLoop.cpp:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Utilities.h:

Canonical link: https://commits.webkit.org/255576@main
